File: python/voice/wake_word.py
#!/usr/bin/env python3
"""
wake_word.py — Wake word detection sidecar.

Streams mic audio through Whisper in 2-second chunks.
Outputs newline-delimited JSON to stdout:
  {"transcript": "hey edith turn on lights", "confidence": 0.9}

Environment:
  WAKE_PHRASE — the wake phrase to detect (default: "hey edith")

Privacy: all audio processing is local. No audio leaves the device.

Fallback: if whisper/sounddevice are not installed, runs in stdin relay mode
(reads lines from stdin and re-emits them as JSON) — useful for testing.
"""
import sys
import json
import os

WAKE_PHRASE = os.environ.get("WAKE_PHRASE", "hey edith").lower()


def emit(transcript: str, confidence: float = 1.0) -> None:
    """Emit a transcript event to stdout as newline-delimited JSON."""
    print(json.dumps({"transcript": transcript, "confidence": confidence}), flush=True)


def run_whisper_mode() -> None:
    """Run with real Whisper + sounddevice for live mic detection."""
    import whisper  # type: ignore[import]
    import sounddevice as sd  # type: ignore[import]
    import numpy as np  # type: ignore[import]

    model = whisper.load_model("tiny")
    SAMPLE_RATE = 16000
    CHUNK_SECONDS = 2

    sys.stderr.write("[wake_word] Whisper tiny loaded — listening for: " + WAKE_PHRASE + "\n")
    sys.stderr.flush()

    while True:
        audio = sd.rec(
            int(CHUNK_SECONDS * SAMPLE_RATE),
            samplerate=SAMPLE_RATE,
            channels=1,
            dtype="float32",
        )
        sd.wait()
        audio_flat = audio.flatten()

        result = model.transcribe(audio_flat, language=None, fp16=False)
        transcript = result.get("text", "").strip()
        if transcript:
            emit(transcript, 0.9)


def run_stdin_relay_mode() -> None:
    """Fallback: relay stdin lines as transcript events (for testing/dev)."""
    sys.stderr.write("[wake_word] whisper/sounddevice not available — stdin relay mode\n")
    sys.stderr.flush()
    for line in sys.stdin:
        line = line.strip()
        if line:
            emit(line)


def main() -> None:
    try:
        run_whisper_mode()
    except ImportError:
        run_stdin_relay_mode()
    except KeyboardInterrupt:
        sys.stderr.write("[wake_word] interrupted\n")
        sys.stderr.flush()


if __name__ == "__main__":
    main()

File: src/voice/wake-word.ts
/**
 * @file wake-word.ts
 * @description Wake word detector — bridges to Python sidecar via child_process stdin/stdout.
 *
 * ARCHITECTURE / INTEGRATION:
 *   Spawns python/voice/wake_word.py which streams mic audio through Whisper in 2s chunks.
 *   Communicates via newline-delimited JSON: { "transcript": "...", "confidence": 0.9 }
 *   Privacy: all audio processing local. No audio leaves device.
 *   Falls back gracefully if Python sidecar unavailable (manual mode: call processTranscript()).
 *   Integrates with always-on.ts for full voice pipeline activation on detection.
 */
import { EventEmitter } from 'node:events'
import { createLogger } from '../logger.js'
import config from '../config.js'

const log = createLogger('voice.wake-word')

/** Wake phrase detection event payload. */
export interface WakeDetectedEvent {
  /** Full transcript that triggered detection. */
  transcript: string
  /** Which wake phrase was matched. */
  phrase?: string
  /** Confidence from sidecar (1.0 if from manual processTranscript call). */
  confidence?: number
}

/** JSON message from the Python sidecar. */
interface SidecarMessage {
  transcript: string
  confidence: number
}

class WakeWordDetector extends EventEmitter {
  private isListening = false
  private sidecarProcess: import('node:child_process').ChildProcess | null = null
  private buffer = ''

  /** All wake phrases to detect (including common misheards). */
  readonly wakePhrases: string[]

  constructor() {
    super()
    this.wakePhrases = [
      (config.WAKE_WORD_PHRASE ?? 'hey edith').toLowerCase(),
      'edith',
      'hey edit', // common mishearing
      'hay edith',
    ]
  }

  /**
   * Start listening for wake word.
   * Attempts to start Python sidecar; falls back to manual mode on failure.
   */
  async start(): Promise<void> {
    if (this.isListening) return
    this.isListening = true

    try {
      await this.startSidecar()
    } catch (err) {
      log.warn('wake word sidecar unavailable — manual mode only', { err })
      // In manual mode, processTranscript() must be called externally
    }

    log.info('wake word detector started', { phrases: this.wakePhrases })
  }

  /** Attempt to spawn the Python Whisper sidecar. */
  private async startSidecar(): Promise<void> {
    const { spawn } = await import('node:child_process')
    const { existsSync } = await import('node:fs')
    const { resolve } = await import('node:path')

    const sidecarPath = resolve(process.cwd(), 'python', 'voice', 'wake_word.py')
    if (!existsSync(sidecarPath)) {
      throw new Error(`Wake word sidecar not found: ${sidecarPath}`)
    }

    this.sidecarProcess = spawn('python', [sidecarPath], {
      stdio: ['pipe', 'pipe', 'pipe'],
      env: { ...process.env, WAKE_PHRASE: this.wakePhrases[0] },
    })

    this.sidecarProcess.stdout?.setEncoding('utf8')
    this.sidecarProcess.stdout?.on('data', (chunk: string) => {
      this.buffer += chunk
      const lines = this.buffer.split('\n')
      this.buffer = lines.pop() ?? ''
      for (const line of lines) {
        if (!line.trim()) continue
        try {
          const event = JSON.parse(line) as SidecarMessage
          this.processTranscript(event.transcript, event.confidence)
        } catch {
          // Non-JSON output from sidecar (log lines etc.) — ignore
        }
      }
    })

    this.sidecarProcess.stderr?.on('data', (data: string) => {
      log.debug('wake word sidecar:', { msg: data.trim() })
    })

    this.sidecarProcess.on('exit', (code) => {
      if (this.isListening) {
        log.warn('wake word sidecar exited unexpectedly', { code })
        this.sidecarProcess = null
      }
    })

    log.info('wake word sidecar started', { pid: this.sidecarProcess.pid })
  }

  /**
   * Stop the detector and kill sidecar process.
   */
  stop(): void {
    this.isListening = false
    if (this.sidecarProcess) {
      this.sidecarProcess.kill('SIGTERM')
      this.sidecarProcess = null
    }
    log.info('wake word detector stopped')
  }

  /**
   * Process a transcript chunk — emits 'detected' if wake phrase found.
   * Called internally by sidecar bridge or externally by STT pipeline.
   * @param transcript - Text to check for wake phrase
   * @param confidence - Confidence score (default 1.0 for manual calls)
   */
  processTranscript(transcript: string, confidence = 1.0): void {
    if (!this.isListening) return
    const lower = transcript.toLowerCase()
    const matched = this.wakePhrases.find((phrase) => lower.includes(phrase))
    if (matched) {
      log.info('wake phrase detected', { phrase: matched, confidence })
      this.emit('detected', {
        transcript,
        phrase: matched,
        confidence,
      } satisfies WakeDetectedEvent)
    }
  }

  /**
   * Check if a transcribed text contains a wake phrase.
   * @param transcript - Text to check
   * @returns True if any wake phrase is found
   */
  containsWakePhrase(transcript: string): boolean {
    const lower = transcript.toLowerCase()
    return this.wakePhrases.some((phrase) => lower.includes(phrase))
  }

  /** Whether the detector is currently active. */
  get active(): boolean {
    return this.isListening
  }
}

/** Singleton wake word detector. */
export const wakeWordDetector = new WakeWordDetector()
